### ISSUE TITLE
Add print handler implementation to LCLCefClient

### ISSUE THREAD

**Issue by dliw** [dliw/fpCEF3#40]
- [ ] TCefPrintHandlerOwn
- [ ] extend LCLCefClient example

**Comment by lucianomartini** [comment]
The pdf printing function runs without any error by following this code: http://wiki.freepascal.org/fpCEF3#Creating_a_pdf , but does not generate the pdf. I think is the same case!

**Comment by dliw** [comment]
>By the way, there is some form to get the html of the site loaded and save to a file?

Please post unrelated questions as new issues or ask in the lazarus forum.

You can use `Browser.MainFrame.GetSource` to get the html source (of the frame) as string.

**Comment by lucianomartini** [comment]
Sorry =[

**Comment by dliw** [comment]
Yes indeed. One needs to implement the print handler in any case.

**Comment by lucianomartini** [comment]
Awaiting the example eagerly!

**Comment by dliw** [comment]
The implementation of the print handler should be rather similar to the dialog handler. Still it is more difficult as the fpCEF class abstraction is missing `TCefPrintHandlerOwn` at the moment.
I don't know, when I will find time to complete this.

**Comment by lucianomartini** [comment]
Don't worry. First thank you for the download implementation, i think this will save my life. I am very happy with the fact that you will implement this two in your project examples!

**Comment by dliw** [comment]
The latest fpCEF version now has a minimal print handler implementation, so that printing to PDF files is now supported.

Complete printing support on Linux is much more complicated unfortunately. There is a  [reference implementation of  the print handler](https://bitbucket.org/chromiumembedded/cef/src/9928036f638f3a2e1bcccf18c28d940e473f82cb/tests/cefclient/browser/print_handler_gtk.cc) in CEF, but it is limited to GTK.

There is platform independent printing support in Lazarus through the `Printer4Lazarus` package, but it does not support printing PDF files, which is needed for the print handler.
The solution probably is to use CUPS directly. However, Printer4Lazarus does not really expose the CUPS interface and I did not find any other package ready to be used.

Maybe I'll try to complete the implementation sometime, but for now this is just too much out of the scope of fpCEF.

**Comment by lucianomartini** [comment]
This helped a lot (pdf printing can be use to run the command lp and send the pdf to the real printer). About the reference i will read it and try to implement. But i am not a programmer, i am Linux administrator but i am trying to learn something. 

Thank you.

**Comment by dliw** [comment]
>About the reference i will read it and try to implement.

If you want to use gtk, then you could indeed try.
But I think it is easier to use CUPS ([API doc](https://www.cups.org/doc/api-cups.html)) directly - basically calling `cupsPrintFile` in `TCustomPrintHandler.OnPrintJob`. A header translation for CUPS can be found in [cupsdyn.pp](http://svn.freepascal.org/svn/lazarus/trunk/components/printers/unix/) and there is a `TCUPSPrinter` class. However, both cannot be used directly - this may be worth a bug report on the Free Pascal bugtracker.
For the print settings `TPrintDialog` could be used.

>I am just a Linux administrator but i am trying to learn something

Well, you started with a really challenging task ...

>if you discover this first than me please share

Of course I will.

>how can i contribute with the project more

Report bugs and ask questions (as long as the answers are not too obvious). Implementing the printing support I actually found typing errors.
More (or extended) examples are very welcome, too. But in the end most of the "work" for fpCEF is working on issues, bugreports and questions :smile:
